tutorials/basic-topics/sorting/merge-sort.md

```diff
@@ -11,4 +11,504 @@ keywords:
+           if leftArray[i] < rightArray[j]: 
+               nums[k] = leftArray[i] 
+               #incrementing the pointer to the smaller element
+               i+=1 
+           else: 
+               nums[k] = rightArray[j] 
+               j+=1
+           k+=1
+
+       while i < len(leftArray): 
+           nums[k] = leftArray[i] 
+           i+=1
+           k+=1
+
+       while j < len(rightArray): 
+           nums[k] = rightArray[j] 
+           j+=1
+           k+=1
+
+```
+</TabItem>
+<TabItem value="java" label="Java">
+
+```java
+// call mergeSort(nums , 0 ,nums.length -1) in public
+// nums -> ArrayList 
+// l -> starting index 
+// r-> end of the ArrayList
+private void mergeSort(int[] nums, int l, int r) {   
+    if (l >= r) return;
+    int mid = l + (r - l) / 2;
+    mergeSort(nums, l, mid);
+    mergeSort(nums, mid + 1, r);
+    merge(nums, l, r);
+}
+
+private void merge(int[] nums, int l, int r) {
+    int mid = l + (r - l) / 2;
+    int[] tmp = new int[r - l + 1];
+    int i = l, j = mid + 1, k = 0;
+    while (i <= mid || j <= r) {
+        if (i > mid || j <= r && nums[i] > nums[j]) {
+            tmp[k++] = nums[j++];
+        } else {
+            tmp[k++] = nums[i++];
+        }
+    }
+}
+```
+</TabItem>
+</Tabs>
+
+## Bottom Up Approach / Iterative technique
+
+### Algorithm 
+```
+1. So it  starts with an  element in the array 
+    It is an iterative approach and because one item array is always sorted
+ 
+2. Compares two nearby elements to merge into a sorted sub array 
+    Similarly, we then merge the sorted subarrays like we have done in top-down recursive approach (two-pointer approach) 
+
+3. Continues until we have a sorted array 
+```
+<Tabs>
+<TabItem value="java" label="Java">
+
+```java
+class Solution {
+    public List<Integer> sortArray(int[] nums) {
+        List<Integer> res = new ArrayList<>();
+        if (nums == null || nums.length == 0) return res;
+        mergeSort(nums);
+        for (int i : nums) res.add(i);
+        return res;
+    }    
+    // iterative only 
+    private void mergeSort(int[] nums) { 
+        for (int size = 1; size < nums.length; size *= 2) {
+            for (int i = 0; i < nums.length - size; i += 2 * size) {
+                int mid = i + size - 1;
+                int end = Math.min(i + 2 * size - 1, nums.length - 1);
+                merge(nums, i, mid, end);
+            }
+        }
+    }
+
+    private void merge(int[] nums, int l, int mid, int r) {
+        int[] tmp = new int[r - l + 1];
+        int i = l, j = mid + 1, k = 0;
+        while (i <= mid || j <= r) {
+            if (i > mid || j <= r && nums[i] > nums[j]) {
+                tmp[k++] = nums[j++];
+            } else {
+                tmp[k++] = nums[i++];
+            }
+        }
+        System.arraycopy(tmp, 0, nums, l, r - l + 1);
+    }
+}
+```
+</TabItem>
+</Tabs>
+
+**Time Complexity** : Best & Worst & Average is   $O(n \log_{2} n)$
```

Comment:
because?

Reply:
Time complexity is explained later @wingkwong